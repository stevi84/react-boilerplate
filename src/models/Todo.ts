import { Components } from '../apis/api';

export type Todo = Components.Schemas.Todo;

export const getEmptyTodo = (): Todo => ({
  id: 0,
  tsCreate: '',
  tsUpdate: '',
  owner: '',
  dueDate: '',
  description: '',
  completed: false,
});
