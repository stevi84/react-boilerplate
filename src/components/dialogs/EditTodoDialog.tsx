import { useNavigate, useParams } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from '../../reducers/Store';
import { useEffect, useState } from 'react';
import { Todo, getEmptyTodo } from '../../models/Todo';
import { Form, Formik, FormikHelpers } from 'formik';
import * as Yup from 'yup';
import { isIsoDate } from '../../globals/Validations';
import { StringEdit } from '../common/StringEdit';
import { useTranslation } from 'react-i18next';
import { DateEdit } from '../common/DateEdit';
import { BooleanEdit } from '../common/BooleanEdit';
import { createTodo, readTodo, updateTodo } from '../../reducers/TodosReducer';
import { CrudMode } from '../dataTable/DataTableInterfaces';
import { Button, Paper, Stack, Tooltip } from '@mui/material';
import CheckIcon from '@mui/icons-material/Check';
import CloseIcon from '@mui/icons-material/Close';
import { isReadingSelector, isSubmittingSelector } from '../../reducers/ApiCallsReducer';
import { Working } from '../common/Working';
import { MainLayout } from '../common/MainLayout';

export const EditTodoDialog = () => {
  const isReading: boolean = useAppSelector(isReadingSelector);
  const isSubmitting: boolean = useAppSelector(isSubmittingSelector);
  const [todo, setTodo] = useState<Todo>(getEmptyTodo());
  const { id } = useParams();
  const { t } = useTranslation();
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  const mode: CrudMode | undefined = id === 'new' ? CrudMode.CREATE : /^\d+$/.test(id!) ? CrudMode.UPDATE : undefined;
  useEffect(() => {
    (async () => {
      if (mode === CrudMode.UPDATE) {
        setTodo(await readTodo(Number(id), dispatch));
      }
    })();
  }, [dispatch, id, mode]);

  const save = (values: Todo, form: FormikHelpers<Todo>) => {
    switch (mode) {
      case CrudMode.CREATE:
        dispatch(createTodo(values));
        break;
      case CrudMode.UPDATE:
        dispatch(updateTodo(values));
    }
    doCancel(form)();
  };

  const doCancel = (form: FormikHelpers<Todo>) => () => {
    form.resetForm({ values: getEmptyTodo() });
    navigate('/todos');
  };

  return (
    <MainLayout>
      <Paper sx={{ margin: 1 }}>
        {!isReading && !isSubmitting && (
          <Formik<Todo>
            initialValues={todo}
            enableReinitialize={true}
            validationSchema={Yup.object({
              owner: Yup.string().required(t('error_empty')),
              dueDate: Yup.string().required(t('error_empty')).test('isIsoDate', t('error_date'), isIsoDate),
              description: Yup.string().required(t('error_empty')),
              completed: Yup.boolean().required(t('error_empty')),
            })}
            onSubmit={save}
          >
            {(form) => (
              <Form>
                <Stack direction={'column'}>
                  <StringEdit id="owner" name="owner" label={t('owner')} />
                  <DateEdit id="dueDate" name="dueDate" label={t('dueDate')} />
                  <StringEdit id="description" name="description" label={t('description')} />
                  <BooleanEdit id="completed" name="completed" label={t('completed')} />
                  <Stack direction={'row'}>
                    <Tooltip
                      title={
                        !form.dirty
                          ? t('tooltip_no_changes')
                          : !form.isValid
                          ? t('tooltip_invalid')
                          : isSubmitting
                          ? t('tooltip_submitting')
                          : t('save')
                      }
                    >
                      <span>
                        <Button
                          id="button-save"
                          variant="contained"
                          color="primary"
                          onClick={form.submitForm}
                          disabled={!form.dirty || !form.isValid || isSubmitting}
                          startIcon={<CheckIcon />}
                          autoFocus={true}
                          sx={{ margin: 1 }}
                        >
                          {t('save')}
                        </Button>
                      </span>
                    </Tooltip>
                    <Tooltip title={t('cancel')}>
                      <span>
                        <Button
                          id="button-cancel"
                          variant="outlined"
                          color="primary"
                          onClick={doCancel(form)}
                          startIcon={<CloseIcon />}
                          sx={{ margin: 1 }}
                        >
                          {t('cancel')}
                        </Button>
                      </span>
                    </Tooltip>
                  </Stack>
                </Stack>
              </Form>
            )}
          </Formik>
        )}
        {(isReading || isSubmitting) && <Working isReading isSubmitting />}
      </Paper>
    </MainLayout>
  );
};
