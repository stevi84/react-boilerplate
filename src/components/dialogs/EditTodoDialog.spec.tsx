import { describe, it, expect, vi } from 'vitest';
import { ShallowRenderer, createRenderer } from 'react-test-renderer/shallow';
import { fireEvent, render, screen } from '@testing-library/react';
import { EditTodoDialog } from './EditTodoDialog';
import { todo1 } from '../../../test/data/Todo';
// import { createTodo, updateTodo } from '../../reducers/TodosReducer';
// import { getEmptyTodo } from '../../models/Todo';

vi.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (str: string) => str,
    i18n: {
      language: 'de',
      changeLanguage: vi.fn(),
    },
  }),
}));

const { navigateMock, useParamsMock } = vi.hoisted(() => ({
  navigateMock: vi.fn(),
  useParamsMock: vi.fn(),
}));
vi.mock('react-router-dom', () => ({
  useNavigate: () => navigateMock,
  useParams: useParamsMock,
}));

vi.mock('../../hooks/UseNotifier', () => ({ useNotifier: vi.fn() }));

const { useAppSelectorMock, dispatchMock } = vi.hoisted(() => ({
  useAppSelectorMock: vi.fn(),
  dispatchMock: vi.fn(),
}));
vi.mock('../../reducers/Store', () => ({
  useAppSelector: useAppSelectorMock,
  useAppDispatch: () => dispatchMock,
}));

const { readTodoMock } = vi.hoisted(() => ({ readTodoMock: vi.fn() }));
readTodoMock.mockResolvedValue(todo1);
vi.mock('../../reducers/TodosReducer', async () => {
  const mod = await vi.importActual<typeof import('../../reducers/TodosReducer')>('../../reducers/TodosReducer');
  return {
    ...mod,
    readTodo: readTodoMock,
  };
});

describe('EditTodoDialog', () => {
  it('should equal saved snapshot', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 1 });
    const renderer: ShallowRenderer = createRenderer();
    renderer.render(<EditTodoDialog />);
    const tree = renderer.getRenderOutput();
    expect(tree).toMatchSnapshot();
  });

  it('should load todo if updating', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: '1' });
    expect(readTodoMock.mock.calls.length).toEqual(0);
    render(<EditTodoDialog />);
    expect(readTodoMock.mock.calls.length).toEqual(1);
    expect(readTodoMock.mock.calls[0][0]).toEqual(1);
    expect(readTodoMock.mock.calls[0][1]).toEqual('de');
    expect(readTodoMock.mock.calls[0][2]).toEqual(dispatchMock);
  });

  it('should not load todo if creating', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 'new' });
    render(<EditTodoDialog />);
    expect(readTodoMock.mock.calls.length).toEqual(0);
  });

  // it('should dispatch createTodo on saving new todo', () => {
  //   useAppSelectorMock.mockReturnValue(false);
  //   useParamsMock.mockReturnValue({ id: 'new' });
  //   expect(dispatchMock.mock.calls.length).toEqual(0);
  //   render(<EditTodoDialog />);
  //   fireEvent.change(screen.getByLabelText('owner'), { target: { value: 'owner' } });
  //   fireEvent.change(screen.getByLabelText('dueDate'), { target: { value: '06.12.2023' } });
  //   fireEvent.change(screen.getByLabelText('description'), { target: { value: 'description' } });
  //   fireEvent.click(screen.getByText('save'));
  //   expect(dispatchMock.mock.calls.length).toEqual(1);
  //   expect(dispatchMock.mock.calls[0][0]).toEqual(createTodo(getEmptyTodo(), 'de'));
  // });

  // it('should dispatch updateTodo on saving existing todo', () => {
  //   useAppSelectorMock.mockReturnValue(false);
  //   useParamsMock.mockReturnValue({ id: '1' });
  //   expect(dispatchMock.mock.calls.length).toEqual(0);
  //   render(<EditTodoDialog />);
  //   fireEvent.change(screen.getByLabelText('owner'), { target: { value: 'change' } });
  //   fireEvent.click(screen.getByText('save'));
  //   expect(dispatchMock.mock.calls.length).toEqual(1);
  //   expect(dispatchMock.mock.calls[0][0]).toEqual(updateTodo(getEmptyTodo(), 'de'));
  // });

  it('should navigate to todos on cancel', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: '1' });
    expect(navigateMock.mock.calls.length).toEqual(0);
    render(<EditTodoDialog />);
    fireEvent.click(screen.getByText('cancel'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/todos');
  });

  it('should show dialog if not reading or submitting', () => {
    useAppSelectorMock.mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 1 });
    render(<EditTodoDialog />);
    expect(screen.getByLabelText('owner')).toBeTruthy();
    expect(screen.queryByRole('progressbar')).toBeFalsy();
  });

  it('should show working if reading', () => {
    useAppSelectorMock.mockReturnValueOnce(true).mockReturnValue(false);
    useParamsMock.mockReturnValue({ id: 1 });
    render(<EditTodoDialog />);
    expect(screen.queryByLabelText('owner')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });

  it('should show working if submitting', () => {
    useAppSelectorMock.mockReturnValueOnce(false).mockReturnValue(true);
    useParamsMock.mockReturnValue({ id: 1 });
    render(<EditTodoDialog />);
    expect(screen.queryByLabelText('owner')).toBeFalsy();
    expect(screen.getByRole('progressbar')).toBeTruthy();
  });
});
