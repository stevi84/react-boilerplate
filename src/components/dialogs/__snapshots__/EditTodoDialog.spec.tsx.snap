// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditTodoDialog > should equal saved snapshot 1`] = `
<MainLayout>
  <ForwardRef(Paper)
    sx={
      {
        "margin": 1,
      }
    }
  >
    <Formik
      enableReinitialize={true}
      initialValues={
        {
          "completed": false,
          "description": "",
          "dueDate": "",
          "id": 0,
          "owner": "",
          "tsCreate": "",
          "tsUpdate": "",
        }
      }
      onSubmit={[Function]}
      validationSchema={
        ObjectSchema {
          "_blacklist": Set {},
          "_excludedEdges": [],
          "_mutate": undefined,
          "_nodes": [
            "completed",
            "description",
            "dueDate",
            "owner",
          ],
          "_sortErrors": [Function],
          "_typeCheck": [Function],
          "_whitelist": Set {},
          "conditions": [],
          "deps": [],
          "exclusiveTests": {},
          "fields": {
            "completed": BooleanSchema {
              "_blacklist": Set {},
              "_mutate": undefined,
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {},
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [],
              "transforms": [
                [Function],
              ],
              "type": "boolean",
            },
            "description": StringSchema {
              "_blacklist": Set {},
              "_mutate": undefined,
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
            "dueDate": StringSchema {
              "_blacklist": Set {},
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "isIsoDate": false,
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
            "owner": StringSchema {
              "_blacklist": Set {},
              "_mutate": undefined,
              "_typeCheck": [Function],
              "_whitelist": Set {},
              "conditions": [],
              "deps": [],
              "exclusiveTests": {
                "required": false,
              },
              "internalTests": {
                "nullable": [Function],
                "optionality": [Function],
                "typeError": [Function],
              },
              "spec": {
                "abortEarly": true,
                "coerce": true,
                "disableStackTrace": false,
                "nullable": false,
                "optional": false,
                "recursive": true,
                "strict": false,
                "strip": false,
              },
              "tests": [
                [Function],
              ],
              "transforms": [
                [Function],
              ],
              "type": "string",
            },
          },
          "internalTests": {
            "nullable": [Function],
            "typeError": [Function],
          },
          "spec": {
            "abortEarly": true,
            "coerce": true,
            "disableStackTrace": false,
            "nullable": false,
            "optional": true,
            "recursive": true,
            "strict": false,
            "strip": false,
          },
          "tests": [],
          "transforms": [],
          "type": "object",
        }
      }
    >
      [Function]
    </Formik>
  </ForwardRef(Paper)>
</MainLayout>
`;
